esapi/esapi/esys_tr.c

```diff
@@ -0,0 +1,379 @@
+    return_if_error(r, "Unmarshal resource object");
+    return TSS2_RC_SUCCESS;
+}
+
+/** Start syncrounous creation of an ESYS_TR object from TPM metadata.
+ *
+ * This function starts the asynchronous retrieval of metadata from the TPM in
+ * order to create a new ESYS_TR object.
+ * @see Esys_TR_FromTPMPublic for more infomration
+ */
+TSS2_RC
+Esys_TR_FromTPMPublic_Async(ESYS_CONTEXT * esys_context,
+                            TPM2_HANDLE tpm_handle,
+                            ESYS_TR shandle1,
+                            ESYS_TR shandle2, ESYS_TR shandle3)
+{
+    TSS2_RC r;
+    ESYS_TR esys_handle = esys_context->esys_handle_cnt++;
+    RSRC_NODE_T *esysHandleNode = NULL;
+    r = esys_CreateResourceObject(esys_context, esys_handle, &esysHandleNode);
+    goto_if_error(r, "Error create resource", error_cleanup);
+    esysHandleNode->rsrc.handle = tpm_handle;
+    esys_context->esys_handle = esys_handle;
+
+    if (tpm_handle >= TPM2_NV_INDEX_FIRST && tpm_handle <= TPM2_NV_INDEX_LAST) {
+        esys_context->in.NV_ReadPublic.nvIndex = esys_handle;
+        r = Esys_NV_ReadPublic_async(esys_context, esys_handle, shandle1,
+                                     shandle2, shandle3);
+        goto_if_error(r, "Error NV_ReadPublic", error_cleanup);
+    } else {
+        esys_context->in.ReadPublic.objectHandle = esys_handle;
+        r = Esys_ReadPublic_async(esys_context, esys_handle, shandle1, shandle2,
+                                  shandle3);
+        goto_if_error(r, "Error ReadPublic", error_cleanup);
+    }
+    return r;
+ error_cleanup:
+    Esys_TR_Close(esys_context, &esys_handle);
+    return r;
+}
+
+/** Finish asyncrounous creation of an ESYS_TR object from TPM metadata.
+ * This function finishes the asynchronous retrieval of metadata from the TPM in
+ * order to create a new ESYS_TR object.
+ * @see Esys_TR_FromTPMPublic for more infomration
+ */
+TSS2_RC
+Esys_TR_FromTPMPublic_Finish(ESYS_CONTEXT * esys_context, ESYS_TR * esys_handle)
+{
+    TSS2_RC r = TSS2_RC_SUCCESS;
+    ESYS_TR objectHandle = esys_context->esys_handle;
+    RSRC_NODE_T *objectHandleNode;
+
+    r = esys_GetResourceObject(esys_context, objectHandle, &objectHandleNode);
+    goto_if_error(r, "get resource", error_cleanup);
+    if (objectHandleNode->rsrc.handle >= TPM2_NV_INDEX_FIRST
+        && objectHandleNode->rsrc.handle <= TPM2_NV_INDEX_LAST) {
+        TPM2B_NV_PUBLIC *nvPublic;
+        TPM2B_NAME *nvName;
+        r = Esys_NV_ReadPublic_finish(esys_context, &nvPublic, &nvName);
+        goto_if_error(r, "Error NV_ReadPublic", error_cleanup);
+        objectHandleNode->rsrc.rsrcType = IESYSC_NV_RSRC;
+        objectHandleNode->rsrc.name = *nvName;
+        objectHandleNode->rsrc.misc.rsrc_nv_pub = *nvPublic;
+        SAFE_FREE(nvPublic);
+        SAFE_FREE(nvName);
+    } else {
+        TPM2B_PUBLIC *public;
+        TPM2B_NAME *name;
+        TPM2B_NAME *qualifiedName;
+        r = Esys_ReadPublic_finish(esys_context, &public, &name,
```

Comment:
`name` is not initialized and it's passed into `Esys_ReadPublic_finish` by reference. If I'm reading this correctly `Esys_ReadPublic_finish` checks the `name` parameter for `NULL` and if it's not null it will be used uninitialized.

Reply:
Same for `qualifiedName`?